### PR TITLE
Full SDXL Model

### PR DESCRIPTION
**Pull request by jazcollins**
This PR contains the full implementation of Stable Diffusion XL (SDXL). SDXL uses two text encoders/tokenizers and also takes crop & size parameters from the dataloader as conditioning - a majority of the changes here are for supporting that. 

**A high-level description of the changes for each file:**

`diffusion/datasets/image_caption.py`
- Added `rand_crop` flag to choose between `LargestCenterSquare` & `RandomCropSquare` - previously only center cropping was supported. This is relevant to SD2 if one might want to train with random cropping, but doesn't apply to SDXL
- Infer whether or not we're doing SDXL training by the `tokenizer_name_or_path`
- Use `RandomCropSquareReturnTransform` for SDXL, which returns the cropping parameters used as well as original image size (for training SDXL with micro-conditioning) and return micro-conditioning as part of the training batch 
- Add option to do micro-conditioning dropout (by setting `microcond_drop_prob` flag). This is not discussed in the SDXL paper but is reflected in [Stability AI's implementation](https://github.com/Stability-AI/generative-models/blob/477d8b9a7730d9b2e92b326a770c0420d00308c9/sgm/modules/encoders/modules.py#L151-L160)
- Small changes necessary for using `SDXLTokenizer`

`diffusion/datasets/laion/transforms.py`
- Implementation of random cropping
- `RandomCropSquare` (does random crop only) and `RandomCropSquareReturnTransform` (does random crop and returns crop params)

`diffusion/models/layers.py`
- Contains attention processor classes for QKV clipping
- Contains `zero_module` function used in SDXL init

`diffusion/models/models.py`
- Add option to do QKV clipping for both SD2 and SDXL (`clip_qkv` argument)
- For SDXL, instantiate `SDXLTokenizer` and `SDXLTextEncoder` which contain the two tokenizers/text encoders but mostly can be used as if they are one tokenizer/text encoder
- For SDXL, do zero init trick

`diffusion/models/stable_diffusion.py`
- Pass `sdxl` flag to StableDiffusion to indicate if we are training an SDXL model
- Set appropriate `latent_scale` for SD2 vs. SDXL
- Extract `pooled_conditioning` from SDXL text encoder, which is used in micro-conditioning
- Construct micro-conditioning dict and pass it to the UNet forward call
- In StableDiffusion.generate(...), allow user to pass `crop_params` and `size_params` for SDXL micro-conditioning, otherwise set them to reasonable default values
- In StableDiffusion.generate(...), new flag called `zero_out_negative_prompt` that zero's out the negative prompt if it is empty (rather than tokenizing and encoding the empty string). This was added to match the behavior of the diffusers [StableDiffusionXLPipeline](https://huggingface.co/docs/diffusers/api/pipelines/stable_diffusion/stable_diffusion_xl) and in general this just seems like a good thing to do. Note: I set the default value to be `True`, so this means previously made generations (e.g. with SD2) will look different despite using the same prompt/seed. Obviously can set it to `False` to match previous results.

`diffusion/callbacks/log_diffusion_images.py`
- Edits to support multiple tokenizers

`setup.py`
- Pin diffusers version to 0.21.0 - this is necessary because attention processor implementations are tied to this version

**There are a few remaining things I'd like to add, but this is already a big enough PR. I will add the following once this one is merged in:**
- Zero-ing out dropped captions rather than tokenizing/encoding the empty string during training (this can apply to SD2 training as well)
- Adding a `CenterCropSquareReturnTransform` transformation that can be used for COCO eval. Currently with SDXL training we do random crop for the eval dataset as well
- Allow user to pass different prompts to the different text encoders in SDXL inference

### CODE REVIEW

File: diffusion/models/stable_diffusion.py [hunk 1]
```diff
@@ -320,10 +376,16 @@ def generate(
                 Default: `3.0`.
             num_images_per_prompt (int): The number of images to generate per prompt.
                  Default: `1`.
-            progress_bar (bool): Wether to use the tqdm progress bar during generation.
+            progress_bar (bool): Whether to use the tqdm progress bar during generation.
                 Default: `True`.
             seed (int): Random seed to use for generation. Set a seed for reproducible generation.
                 Default: `None`.
+            zero_out_negative_prompt (bool): Whether or not to zero out negative prompt if it is
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply diffusion/models/stable_diffusion.py:383]: For the forward pass, I'd say do it by default, no argument.

File: diffusion/models/stable_diffusion.py [hunk 2]
```diff
@@ -366,6 +440,23 @@ def generate(
...
+            if not crop_params:
+                crop_params = [0., 0.]
+            if not size_params:
+                size_params = [width, height]
+            cond_original_size = torch.tensor([[width, height]]).repeat(pooled_embeddings.shape[0],
+                                                                        1).to(device).float()
+            cond_crops_coords_top_left = torch.tensor([crop_params]).repeat(pooled_embeddings.shape[0],
+                                                                            1).to(device).float()
+            cond_target_size = torch.tensor([size_params]).repeat(pooled_embeddings.shape[0], 1).to(device).float()
+            add_time_ids = torch.cat([cond_original_size, cond_crops_coords_top_left, cond_target_size], dim=1).float()
```
Comment (Landanjs) [inline diffusion/models/stable_diffusion.py:455]: What do you think of (not necessary):
```suggestion
            add_time_ids = torch.tensor([[width, height, *crop_params, *size_params]], dtype=torch.float, device=device)
            add_time_ids = add_time_ids.repeat(pooled_embeddings.shape[0], 1)
```

**Review by Landanjs — COMMENTED** [review]
This is great work!! It seems really difficult to manage the SDXL specific features and the SD2 features in one model, but you did the best organization possible. The few possible refactors may not be that much better. I proposed a few suggestions, but up to your discretion if you think they are good / worth the effort.

The only possible bug I noticed was in `log_diffusion_images.py`.

One general proposal: there are a few if-statements due to slight differences in the SDXL and HF tokenizers. If you're up for it, I think it would clean up a bit of code if could make them more similar. I think this requires two things:
1. Having a `max_length` argument in the SDXL Tokenizers and in the code set max length by `max_length = None if self.sdxl else self.tokenizer.model_max_length`.
2. Have the SDXL tokenizer return of dictionary with the key `input_ids`

File: diffusion/models/stable_diffusion.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jazcollins) [reply diffusion/models/stable_diffusion.py:383]: Will do in next PR!

File: diffusion/callbacks/log_diffusion_images.py [hunk 3]
```diff
@@ -62,14 +62,25 @@ def eval_batch_end(self, state: State, logger: Logger):
...
-                self.tokenized_prompts = torch.cat(tokenized_prompts)
-            self.tokenized_prompts = self.tokenized_prompts.to(state.batch[self.text_key].device)
+            if model.sdxl:
+                if self.tokenized_prompts is None:
+                    tokenized_prompts = [
+                        model.tokenizer(p, padding='max_length', truncation=True, return_tensors='pt',
+                                        input_ids=True)  # type: ignore
+                        for p in self.prompts
+                    ]
+                    self.tokenized_prompts = [torch.cat(tokenized_prompts[0]), torch.cat(tokenized_prompts[1])]
```
[the remark this answers is not in the document]
Comment (jazcollins) [reply diffusion/callbacks/log_diffusion_images.py]: fixed in https://github.com/mosaicml/diffusion/pull/67/commits/505b850711d754840e82f1ec68393f0a626e7f69

File: diffusion/callbacks/log_diffusion_images.py [hunk 4]
```diff
@@ -62,14 +62,25 @@ def eval_batch_end(self, state: State, logger: Logger):
...
-                                    return_tensors='pt')['input_ids']  # type: ignore
-                    for p in self.prompts
-                ]
-                self.tokenized_prompts = torch.cat(tokenized_prompts)
-            self.tokenized_prompts = self.tokenized_prompts.to(state.batch[self.text_key].device)
+            if model.sdxl:
+                if self.tokenized_prompts is None:
+                    tokenized_prompts = [
+                        model.tokenizer(p, padding='max_length', truncation=True, return_tensors='pt',
+                                        input_ids=True)  # type: ignore
```
[the remark this answers is not in the document]
Comment (jazcollins) [reply diffusion/callbacks/log_diffusion_images.py]: done in https://github.com/mosaicml/diffusion/pull/67/commits/054d1ef454d62220b23b348b7d4296fed52b4dbb!

**Review by Landanjs — APPROVED** [review]
Amazing!! Thanks for all the fixes!

File: diffusion/models/models.py [hunk 5]
```diff
@@ -181,39 +199,34 @@ def stable_diffusion_xl(
...
+            # proj_out in attention blocks
+            if name.endswith('to_out.0'):
+                layer = zero_module(layer)
+        # Last conv block out projection
+        unet.conv_out = zero_module(unet.conv_out)
+
+    torch_dtype = torch.float16 if encode_latents_in_fp16 else None
+    try:
+        vae = AutoencoderKL.from_pretrained(vae_model_name, subfolder='vae', torch_dtype=torch_dtype)
+    except:  # for handling SDXL vae fp16 fixed checkpoint
```
Comment (Skylion007) [inline diffusion/models/models.py:221]: The blanket except isn't great here. We should probably qualify the exception type here.

File: diffusion/models/models.py [hunk 6]
```diff
@@ -231,12 +244,22 @@ def stable_diffusion_xl(
...
         if is_xformers_installed:
             model.unet.enable_xformers_memory_efficient_attention()
             model.vae.enable_xformers_memory_efficient_attention()
+
+    if clip_qkv is not None:
+        if is_xformers_installed:
+            attn_processor = ClippedXFormersAttnProcessor(clip_val=clip_qkv)
+        else:
+            attn_processor = ClippedAttnProcessor2_0(clip_val=clip_qkv)
+        log.info('Using %s with clip_val %.1f' % (attn_processor.__class__, clip_qkv))
```
Comment (Skylion007) [inline diffusion/models/models.py:260]: Remove the '%' sign, logger can automatically generate the strings by itself.

File: diffusion/models/layers.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+        hidden_states = hidden_states.to(query.dtype)
+        hidden_states = attn.batch_to_head_dim(hidden_states)
+
+        # linear proj
+        hidden_states = attn.to_out[0](hidden_states, scale=scale)
+        # dropout
+        hidden_states = attn.to_out[1](hidden_states)
+
+        if input_ndim == 4:
+            assert channel
```
Comment (Skylion007) [inline diffusion/models/layers.py:214]: print the value of the assert, so we know why it's False (None, 0, etc.)
```suggestion
            assert channel, f"{channel}"
```

**Review by Skylion007 — COMMENTED** [review]
Great PR, just some minor nits

**Comment by felixdae** [comment]
why we need to clip qkv
